wallet, spkm: Move key management from DescriptorScriptPubKeyMan to wallet level KeyManager

achow101 opened the pull request:
This PR changes DescriptorScriptPubKeyMan to no longer handle relevant keys directly. Instead all keys for all DescriptorSPKMs will be handled by a new `KeyManager` class which exists within `CWallet` (a reference to it is passed to each DescriptorSPKM). This allows us to have a concept of a wallet HD key for descriptor wallets. This makes it easier to add new single key descriptors that use the same HD master key as the rest of the autogenerated descriptors (e.g. for taproot). Multisigs will also be easier as an xpub belonging to the wallet can be exported without needing to do weird things like descriptor introspection and guessing about which descriptor's key to use.

`KeyManager` is a class which handles all of the keys for DescriptorSPKMs. It contains the maps that hold the keys, deals with writing those keys to disk, and handles their encryption. Encryption keys are still managed by `CWallet` but provided to `KeyManager` through the `WalletStorage` interface. Signing is still done through `DescriptorScriptPubKeyMan::SignTransaction` however this will fetch the keys from `KeyManager` rather than storing keys in the DescriptorSPKM to be used.

This change is backwards compatible. Although `KeyManager` writes and uses keys in new `keyman_key` and `keyman_ckey` records, it will still write keys for each descriptor in `walletdescriptorkey` and `walletdescriptorckey` records. This allows a descriptor wallet created using this change to be opened by 22.0 and 0.21. Additionally, wallets created with older software will automatically be upgraded to using the `KeyManager` at first loading. This is done in the background and does not require any user interaction (i.e. no passphrase required).

achow101 committed:
descriptor: Be able to get the pubkeys involved in a descriptor

achow101 committed:
wallet: Add KeyManager class

achow101 committed:
moveonly: move WalletStorage to separate file

achow101 committed:
descspkm: Add KeyManager to DescriptorScriptPubKeyMan and use for keys

achow101 committed:
descspkm: Encrypt with KeyManager instead of direct map access

achow101 committed:
descspkm: Track CKeyIDs of our keys

When DescriptorScriptPubKeyMan no longer manages its keys, it still
needs to know the IDs of its keys.

achow101 committed:
descspkm: Use KeyManager::LoadKey and LoadCryptedKey when loading

achow101 committed:
descspkm: Replace GetKeys with KeyManager::GetKeys

achow101 committed:
descspkm: Replace HavePrivateKeys with KeyManager::HavePrivateKeys()

achow101 committed:
keyman: Make some members private

achow101 committed:
wallet: Have KeyManager in CWallet rather than DescriptorScriptPubKeyMan

achow101 committed:
walletdb: Allow duplicate descriptor keys

If a descriptor (crypted) key is being written and one already exists,
make sure that the one being written and the one already on disk
match each other.

achow101 committed:
walletdb: Load keys into KeyManager directly

achow101 committed:
wallet: Add flag for using KeyManager

KeyManager will be a backwards compatible background upgrade to
descriptor wallets. A flag indicating that the upgrade has occurred is
added so that the upgrade (not yet implemented) will only happen once.

achow101 committed:
wallet: Use KeyManager to generate master key

achow101 committed:
walletdb: Implement upgrading a wallet to use KeyManager

achow101 committed:
descspkm: Remove unneeded key loading

Key management will be done entirely by KeyManager, so
DescriptorScriptPubKeyMan does not need key loading functions.

achow101 committed:
rpc: Add getxpub command

achow101 committed:
walletdb: Add HDKey records

achow101 committed:
walletdb: Add WriteKeyManKey and WriteCryptedKeyManKey

These functions write new key records for keys handled by a KeyManager

Sjors commented:
Concept ACK. This will make #22341 a _lot_ easier.

Sjors commented on src/wallet/walletdb.cpp:
```diff
@@ -1085,7 +1085,97 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
+        std::map<CExtPubKey, std::pair<std::map<std::string, int>, uint64_t>> descs_keys;
+        std::map<std::string, int> tmpl = {{"pkh(", 0}, {"sh(wpkh(", 0}, {"wpkh(", 0}};
+
+        // Find xpubs used in pkh(), sh(wpkh()), and wpkh() descriptors
```
1c76661: maybe clarify that we're dealing with the xpub at the root level, not the BIP44/49/84 "account" level.

achow101 replied:
Done

Sjors reviewed (commented):
I rebased #22341 on top of this.

Some initial review remarks:

<details><summary>wallet_transactiontime_rescan.py fails consistently for me</summary>

```
2021-11-05T14:52:37.129000Z TestFramework (INFO): Restore user wallet on another node without rescan
2021-11-05T14:52:37.154000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/Users/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "test/functional/wallet_transactiontime_rescan.py", line 134, in run_test
    assert_equal(restorewo_wallet.getbalance(), 0)
  File "/Users/sjors/dev/bitcoin/test/functional/test_framework/util.py", line 50, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(1.00000000 == 0)
```
</details>

<details>
  <summary>Some thread safety warnings.</summary>

```
wallet/keyman.cpp:241:5: warning: calling function 'AssertLockHeldInternal<AnnotatedMixin<std::recursive_mutex>>' requires holding mutex 'cs_keyman' exclusively [-Wthread-safety-analysis]
    AssertLockHeld(cs_keyman);
    ^
./sync.h:83:28: note: expanded from macro 'AssertLockHeld'
#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
                           ^
wallet/keyman.cpp:244:41: warning: reading variable 'm_map_crypted_keys' requires holding mutex 'cs_keyman' [-Wthread-safety-analysis]
        for (const auto& [id, key_pair] : m_map_crypted_keys) {
                                        ^
wallet/keyman.cpp:244:41: warning: reading variable 'm_map_crypted_keys' requires holding mutex 'cs_keyman' [-Wthread-safety-analysis]
wallet/keyman.cpp:253:12: warning: reading variable 'm_map_keys' requires holding mutex 'cs_keyman' [-Wthread-safety-analysis]
    return m_map_keys;
           ^
wallet/keyman.cpp:264:9: warning: reading variable 'm_map_crypted_keys' requires holding mutex 'cs_keyman' [-Wthread-safety-analysis]
    if (m_map_crypted_keys.count(id) == 0) {
        ^
wallet/keyman.cpp:267:12: warning: reading variable 'm_map_crypted_keys' requires holding mutex 'cs_keyman' [-Wthread-safety-analysis]
    return m_map_crypted_keys.at(id);
           ^
6 warnings generated.

...

wallet/walletdb.cpp:813:38: warning: calling function 'LoadActiveHDKey' requires holding mutex 'pwallet->GetKeyManager().cs_keyman' exclusively [-Wthread-safety-analysis]
            pwallet->GetKeyManager().LoadActiveHDKey(extpub);
                                     ^
wallet/walletdb.cpp:1150:38: warning: calling function 'SetActiveHDKey' requires holding mutex 'pwallet->GetKeyManager().cs_keyman' exclusively [-Wthread-safety-analysis]
            pwallet->GetKeyManager().SetActiveHDKey(best_xpub);

```
</details>

Shameless plug: it would be useful is #19013 was merged first, so this PR can include more wallet backward and forward compatibility tests.

achow101 commented on src/pubkey.h:
```diff
@@ -304,6 +304,7 @@ struct CExtPubKey {
     {
         return !(a == b);
     }
+    bool operator<(const CExtPubKey& other) const { return pubkey < other.pubkey; }
```
Done

achow101 commented on src/wallet/walletdb.h:
```diff
@@ -55,6 +55,7 @@ namespace DBKeys {
 extern const std::string ACENTRY;
 extern const std::string ACTIVEEXTERNALSPK;
 extern const std::string ACTIVEINTERNALSPK;
+extern const std::string ACTIVEHDKEY;
```
Done

achow101 commented on src/wallet/walletdb.h:
```diff
@@ -63,6 +64,7 @@ extern const std::string DEFAULTKEY;
 extern const std::string DESTDATA;
 extern const std::string FLAGS;
 extern const std::string HDCHAIN;
+extern const std::string HDPUBKEY;
```
Done

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -1085,7 +1085,97 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_USES_KEYMAN) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)
+        && last_client <= 220000
```
Done, also changed this to be either the flag is not set, or the last client is <= 229999. This is to handle the upgrade then downgrade and then upgrade again case.

achow101 commented:
Thread safety warnings are fixed, not seeing the wallet_transactiontime_rescan.py failure.

Sjors reviewed (commented):
Some more comments after light-weight code review...

achow101 commented on src/script/descriptor.h:
```diff
@@ -146,6 +146,8 @@ struct Descriptor {
     /** @return The OutputType of the scriptPubKey(s) produced by this descriptor. Or nullopt if indeterminate (multiple or none) */
     virtual std::optional<OutputType> GetOutputType() const = 0;
+
+    virtual void GetPubkeys(std::set<CPubKey>& pubkeys, std::set<CExtPubKey>& ext_pubs) const = 0;
```
Not particularly. I believe originally I wanted the underlying function in `PubkeyProvider` to be `GetPubkey` but that was already taken. The `Root` part came from the fact that `GetPubkey` in `BIP32PubkeyProvider` returns a derived key whereas I needed the root extended key.

achow101 replied:
Done

achow101 commented on src/wallet/walletdb.h:
```diff
@@ -63,8 +64,11 @@ extern const std::string DEFAULTKEY;
 extern const std::string DESTDATA;
 extern const std::string FLAGS;
 extern const std::string HDCHAIN;
+extern const std::string HDPUBKEY; // A HD key, identified by extended pubkey
```
A lot of places in the codebase do not expect `CExtKey`s when fetching private keys, even in places where BIP 32 derivation ends up being done. They instead take `CKey`s and combine them with `CExtPubKey`s to get the necessary `CExtKey`s. We maintain this same paradigm for ease of implementation.

Additionally, having all private keys be universally `CKey`s makes it easier to support non-ranged descriptors. Instead of having to store and fetch different data types depending on whether the descriptor is ranged, and then converting them into the same data type for expansion and signing, we can use the same datatype throughout. When we do need the extended key, it can be reconstructed, but that (currently) happens rarely.

Sjors replied:
Storing both `CExtKey` and `Ckey` (for non-ranged descriptors) could be an approach, but I'm also not sure if that makes the implementation any easier to understand. This is probably only opportunity to break with the past convention, if we want to.

achow101 commented on src/pubkey.h:
```diff
@@ -304,6 +304,7 @@ struct CExtPubKey {
     {
         return !(a == b);
     }
+    bool operator<(const CExtPubKey& other) const { return vchFingerprint < other.vchFingerprint || pubkey < other.pubkey; }
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -185,6 +185,8 @@ struct PubkeyProvider
     /** Derive a private key, if private data is available in arg. */
     virtual bool GetPrivKey(int pos, const SigningProvider& arg, CKey& key) const = 0;
+
+    virtual void GetRootPubkey(std::set<CPubKey>& pubkeys, std::set<CExtPubKey>& ext_pubs) const = 0;
```
Done

Sjors commented:
Not seeing any thread warning anymore, so that's good.

`wallet_transactiontime_rescan.py` only fails for me with configuring `--without-bdb`.

achow101 commented:
> `wallet_transactiontime_rescan.py` only fails for me with configuring `--without-bdb`.

This appears to fail on master as well.

Sjors commented on src/wallet/keyman.cpp:
```diff
@@ -0,0 +1,272 @@
+        CKeyingMaterial secret(key.begin(), key.end());
+        std::vector<unsigned char> crypted_secret;
+        if (!EncryptSecret(master_key, secret, pubkey.GetHash(), crypted_secret)) {
+            return false;
```
Clear `m_map_crypted_keys` first?

achow101 replied:
I think it's fine to not do this. The caller will `assert(false)` in that case anyways.

Sjors commented on src/wallet/keyman.cpp:
```diff
@@ -0,0 +1,272 @@
+            break;
+        }
+        keyPass = true;
+        if (m_decryption_thoroughly_checked)
```
Nit: move `break` inline or add brackets.

achow101 replied:
Done

Sjors reviewed (approved):
utACK 709a917c26bf048de454760c7189278b965ca7f3 modulo `AddDescriptorKeyWithDB` definition

I found it easier to review the PR as a whole rather the migration through individual commits: `git diff HEAD~21 --color-moved=dimmed_zebra`. This is because some of the function bodies aren't moved in single commits.

Looks like you got rid of `AddDescriptorKeyWithDB`, but it's still defined in header. Also the original function contained an assert for not having `WALLET_FLAG_DISABLE_PRIVATE_KEYS` and it would check if the key already existed (`AddKeyInner` does that now).

(most comments below are about existing code)

achow101 commented on src/wallet/keyman.h:
```diff
@@ -0,0 +1,65 @@
+    std::optional<CExtKey> GetActiveHDKey() const EXCLUSIVE_LOCKS_REQUIRED(cs_keyman);
+    std::optional<CExtPubKey> GetActiveHDPubKey() const EXCLUSIVE_LOCKS_REQUIRED(cs_keyman);
+    void SetActiveHDKey(const CExtPubKey& extpub);
+    void LoadActiveHDKey(const CExtPubKey& extpub);
```
Done

achow101 commented on src/wallet/keyman.cpp:
```diff
@@ -0,0 +1,272 @@
+std::map<CKeyID, CKey> KeyManager::GetKeys() const
+{
+    AssertLockHeld(cs_keyman);
+    if (m_storage.HasEncryptionKeys() && !m_storage.IsLocked()) {
```
Added a comment.

achow101 commented:
> Looks like you got rid of `AddDescriptorKeyWithDB`, but it's still defined in header

Fixed

> Also the original function contained an assert for not having `WALLET_FLAG_DISABLE_PRIVATE_KEYS` and it would check if the key already existed (`AddKeyInner` does that now).

Added that back in.

Sjors commented:
re-utACK b1d76a8d91a070668332f06608a20a9d937ba1a4

(the two spurious CI failures should go away after the next rebase)

MarcoFalke commented:
`/usr/include/c++/8/bits/stl_function.h:386:20: error: ambiguous overload for ‘operator<’ (operand types are ‘const CExtPubKey’ and ‘const CExtPubKey’)`

Sjors commented:
re-utACK 49234d0dbc11f91e9ae657602c97b4a0dc0ec75e
(just a rebase, but with 2edf8b89b4 dropped entirely)

Sjors commented:
re-utACK 01a4860 (after #23497 namespace stuff)

Sjors commented:
re-utACK 3992d06c15af5242d955fd7f74e9e089bbb8a166

Sjors commented:
CI is not happy :-(

Sjors commented:
re-utACK  8133ce4da3fbbcbecab4346cbe3cc825f4467ce7

Sjors commented:
Mmm, CI is not running bf104948b42f63d60afee52ae95a8515432eeb1c.

Looks like #25734 triggered the (trivial) rebase (see ffcdac846b)?

Building throws a warning, because #25642 marked it as such.

```
wallet/keyman.cpp:96:9: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
        ext_key->Derive(*ext_key, i);
```

Tested that #22341 still works on top of this.

Sjors commented:
> Is this for a particular commit?

Oops, in my own code :-)

re-utACK e310760da891c217d2f7805978b5eca337f5cdc8

w0xlt commented on src/wallet/walletdb.cpp:
```diff
@@ -943,6 +1080,100 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
+            }
+
+            const std::string desc = w_desc.descriptor->ToString();
+            if (desc.find("pkh(") == 0) {
```
nit:
```suggestion
            if (desc.find("pkh(")) {
```

achow101 replied:
Done

Sjors replied:
Are you sure? `find` returns the position, so this should match `wpkh(…)` but not `pkh()`. The `else if`'s seem wrong for the same reason.

w0xlt replied:
Maybe `desc.rfind("pkh(", 0) == 0` ?

Sjors replied:
I think the original `desc.find("pkh(") == 0` was fine.

achow101 replied:
Ah, yes. Looked at the docs this time. I've changed these all to be `== 0`.

w0xlt commented on src/wallet/rpc/addresses.cpp:
```diff
@@ -786,4 +786,48 @@ RPCHelpMan walletdisplayaddress()
 }
 #endif // ENABLE_EXTERNAL_SIGNER
+
+RPCHelpMan getxpub()
```
Maybe `getxupub` and `listdescriptors` can be combined into a single command?

achow101 replied:
I think it's useful to have them be separate.

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/23417/commits/aa91428a1a017d833bd542c8c99fe53d9fc11d1f

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -296,6 +316,69 @@ bool WalletBatch::EraseLockedUTXO(const COutPoint& output)
+{
+    // hash pubkey/privkey to accelerate wallet load
+    std::vector<unsigned char> key;
+    key.reserve(pubkey.size() + pubkey.size());
```
Done

achow101 commented on src/wallet/keyman.h:
```diff
@@ -0,0 +1,68 @@
+    std::vector<unsigned char> AddCryptedKeyInner(WalletBatch& batch, const CKey& key, const CPubKey& pubkey) EXCLUSIVE_LOCKS_REQUIRED(cs_keyman);
+
+public:
+    mutable RecursiveMutex cs_keyman;
```
It ends up being used recursively.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/23417/commits/aa91428a1a017d833bd542c8c99fe53d9fc11d1f

Sjors commented:
Rebase to e8926b1 looks correct, but see inline issue above.

Sjors commented:
re-utACK 5323e56ffb8762d5d99bb89d1623993ad6db5849

achow101 commented:
Pushed two small changes. The first is to set `WALLET_USES_KEYMAN` for all newly created wallets to avoid any possibility of attempting to upgrade a wallet when it shouldn't. The second is to bump the last client version check because this isn't going to make it to 24.0.

achow101 commented:
On second thought, I've removed the last client check. It doesn't quite make sense to have that.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/23417/commits/93294e30f1f88246c47bab2bd6a091f89a520a68

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2839,6 +2839,9 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
         walletInstance->InitWalletFlags(wallet_creation_flags);
 
+        // Set USES_KEYMAN to prevent auto upgrade from trying to auto upgrade (previously) blank wallets that had imported things
+        walletInstance->SetWalletFlag(WALLET_FLAG_USES_KEYMAN);
```
eb75f8d58ec335e013178d5b94d824cc7739139a: maybe put this under the `if (!walletInstance->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS))` check below?

Sjors reviewed (approved):
re-utACK 6bdd8a6759c9fc93cd19612bfbd39c1abcc4de00

aureleoules commented on src/wallet/keyman.cpp:
```diff
@@ -0,0 +1,278 @@
+        for (const auto& [id, key_pair] : m_map_crypted_keys) {
+            const auto& [pubkey, crypted_secret] = key_pair;
+            CKey key;
+            bool ok = DecryptKey(m_storage.GetEncryptionKey(), crypted_secret, pubkey, key);
+            assert(ok);
```
can inline this assert
```suggestion
            bool ok = Assert(DecryptKey(m_storage.GetEncryptionKey(), crypted_secret, pubkey, key));
```

achow101 replied:
Assertions should not have side effects, hence the separation here.

aureleoules replied:
this is true for `assert` but `Assert` is a macro defined in `util/check.h` that always executes and returns the statement, but doesn't always verify the assertion
https://github.com/bitcoin/bitcoin/blob/100949af0e2551f22c02a73355f2c64710b68ef1/src/util/check.h#L57-L71

achow101 replied:
Even so, it is more confusing to read be inlining as assertions are expected to be checking invariants rather than doing something.

aureleoules replied:
Well the `Assert` macro has been used extensively already in the codebase so I don't think it's confusing anymore.
But this is a small nit so if you prefer it this way it's fine.

Sjors replied:
I also prefer to keep them on separate lines, for easier review since I've gotten used to that convention.

aureleoules commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2024,56 +2024,24 @@ isminetype DescriptorScriptPubKeyMan::IsMine(const CScript& script) const
-        m_map_crypted_keys[pubkey.GetID()] = make_pair(pubkey, crypted_secret);
-        batch->WriteCryptedDescriptorKey(GetID(), pubkey, crypted_secret);
+    for (const CKeyID& id : m_set_stored_keys) {
+        const auto& ckey_pair = m_keyman.GetCryptedKey(id);
+        assert(ckey_pair != std::nullopt);
```
```suggestion
        const auto& ckey_pair = Assert(m_keyman.GetCryptedKey(id));
```

achow101 replied:
Assertions shouldn't have side effects, hence the separation.

aureleoules reviewed (commented):
Left some code review nits, otherwise looks good to me.
ACK 8c5ddf432dceb084be6612b378dacf9211e8cf08.

Though, shouldn't 8c5ddf432dceb084be6612b378dacf9211e8cf08 go in #22341?

achow101 commented on src/wallet/keyman.cpp:
```diff
@@ -0,0 +1,278 @@
+bool KeyManager::HavePrivateKeys() const
+{
+    LOCK(cs_keyman);
+    return m_map_keys.size() > 0 || m_map_crypted_keys.size() > 0;
```
Done

achow101 commented on src/wallet/keyman.cpp:
```diff
@@ -0,0 +1,278 @@
+    return m_map_keys.size() > 0 || m_map_crypted_keys.size() > 0;
+}
+
+const std::optional<std::pair<CPubKey, std::vector<unsigned char>>> KeyManager::GetCryptedKey(const CKeyID& id) const
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3434,53 +3437,41 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
 {
     if (IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
-        auto spk_manager = std::unique_ptr<ScriptPubKeyMan>(new ExternalSignerScriptPubKeyMan(*this, desc));
+        auto spk_manager = std::unique_ptr<ScriptPubKeyMan>(new ExternalSignerScriptPubKeyMan(*this, desc, m_keyman));
```
Done

aureleoules commented on src/wallet/keyman.cpp:
```diff
@@ -0,0 +1,278 @@
+bool KeyManager::HavePrivateKeys() const
+{
+    LOCK(cs_keyman);
+    return m_map_keys.empty() || m_map_crypted_keys.empty();
```
i think you meant 
```suggestion
    return !m_map_keys.empty() || !m_map_crypted_keys.empty();
```

aureleoules replied:
also, maybe `HasPrivateKeys` is better than `HavePrivateKeys` ?

achow101 replied:
Oops, yes.

> also, maybe `HasPrivateKeys` is better than `HavePrivateKeys` ?

I prefer to keep it the same for consistent naming.

Sjors replied:
Yikes. Can you add a test for this?

aureleoules reviewed (approved):
re-ACK 5145d1610202356676e5a031c68d4af0b16a62d9 - minor changes since last review

aureleoules commented on src/wallet/walletdb.cpp:
```diff
@@ -725,14 +810,76 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
+        } else if (strType == DBKeys::HDPUBKEY) {
+            CExtPubKey extpub;
+            std::vector<unsigned char> xpub(BIP32_EXTKEY_SIZE);
+            ssKey >> xpub;
+            extpub.Decode(xpub.data());
+            pwallet->GetKeyManager().LoadHDKey(extpub.pubkey.GetID(), extpub);
+        } else if (strType == DBKeys::KEYMAN_KEY) {
+            CPubKey pubkey;
+            ssKey >> pubkey;
+            if (!pubkey.IsValid())
+            {
+                strErr = "Error reading wallet database: CPubKey corrupt";
+                return false;
+            }
+            CKey key;
+            CPrivKey pkey;
+            uint256 hash;
+
+            wss.nKeys++;
+            ssValue >> pkey;
+            ssValue >> hash;
+
+            // hash pubkey/privkey to accelerate wallet load
+            std::vector<unsigned char> to_hash;
+            to_hash.reserve(pubkey.size() + pkey.size());
+            to_hash.insert(to_hash.end(), pubkey.begin(), pubkey.end());
+            to_hash.insert(to_hash.end(), pkey.begin(), pkey.end());
+
+            if (Hash(to_hash) != hash)
+            {
+                strErr = "Error reading wallet database: CPubKey/CPrivKey corrupt";
+                return false;
+            }
+
+            if (!key.Load(pkey, pubkey, true))
+            {
+                strErr = "Error reading wallet database: CPrivKey corrupt";
+                return false;
+            }
+            pwallet->GetKeyManager().LoadKey(pubkey.GetID(), key);
```
this seems to be duplicated code from the `else if (strType == DBKeys::WALLETDESCRIPTORKEY)` case above, is there a way to re-use the code?

achow101 replied:
Done

achow101 committed:
walletdb: Refactor deserialaization of keys with checksums

This will become shared later.

aureleoules reviewed (approved):
re-ACK bdc59fcf8a5ef473461c43c2a035accd4f89e35d

aureleoules reviewed (approved):
reACK c9af030cd60f1fe4e66ccba585b616c1dcc11a50 - minor rebase, added `KeyManager m_keyman` to `CWallet`

Sjors commented:
re-utACK c9af030cd60f1fe4e66ccba585b616c1dcc11a50

S3RK commented:
Concept ACK. Started code review

achow101 commented:
I had a discussion with @S3RK about this PR last week, and we concluded that this might not be the best approach to tackle the issue that it targets. While it is a complete solution that would probably work well if descriptor wallets had been implemented this way in the first place, it seems like it doesn't work well with having to deal with backwards compatibility and various combinations of upgrading and downgrading that may occur.

Since the goal of this PR is to enable key rotation (via re-enabling `sethdseed`) and the addition of new automatically generated descriptors, this implementation is probably overkill in addition to the backwards compatibility issues that it introduces.

I've opened #26728 which implements a much simpler solution of just having `CWallet` store the master key and any ones that get rotated out. It still enables the things that we want.
